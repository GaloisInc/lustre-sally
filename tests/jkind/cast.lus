-- This file has four properties, all of which are true for K = 0.
-- It stress the casting functions
--   real : int -> real
--   floor : real -> int

node is_int(x : real) returns (ok : bool);
let
  ok = (x = real(floor(x)));
tel;

node main(x : real; y : int) returns ();
var
  ok1, ok2, ok3, ok4 : bool;
let
  ok1 = real(floor(x)) <= x and x < real(floor(x)) + 1.0;
  --%PROPERTY ok1;

  ok2 = (floor(-x) = -floor(x)) = is_int(x);
  --%PROPERTY ok2;

  ok3 = (y div 5 = floor(real(y) / 5.0));
  --%PROPERTY ok3;

  ok4 = (floor(x) div 5 = floor(x / 5.0));
  --%PROPERTY ok4;
tel;
