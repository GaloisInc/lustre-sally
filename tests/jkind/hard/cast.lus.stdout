Core Lustre
===========

node (x_4 : real, y : int)
  returns ()
  assumes ()
  shows (ok4, ok3, ok2, ok1)
let
  ok1 = And(Leq(RealCast(FloorCast(x_4)), x_4), Lt(x_4, Add(RealCast(FloorCast(x_4)), 1.0)))
  x_11 = x_4
  __no_static = Eq(x_11, RealCast(FloorCast(x_11)))
  ok2 = Eq(Eq(FloorCast(Neg(x_4)), Neg(FloorCast(x_4))), __no_static)
  ok3 = Eq(Div(y, 5), FloorCast(Div(RealCast(y), 5.0)))
  ok4 = Eq(Div(FloorCast(x_4), 5), FloorCast(Div(x_4, 5.0)))
tel
Sally Model
===========

(define-state-type
   S
   (
     (ok1 Bool)
     (ok2 Bool)
     (ok3 Bool)
     (ok4 Bool)
     (y Int)
     (|__no_static| Bool)
     (|gal-initializing| Bool)
     (|x:11| Real)
     (|x:4| Real))
   (
     (y Int)
     (|x:4| Real)))
(define-transition-system
   TS
   S
   (= |gal-initializing| true)
   (and
      (= next.|gal-initializing| false)
      (= next.|x:4| input.|x:4|)
      (= next.y input.y)
      (=
         next.ok1
         (and
            (<=
               (to_real
                  (to_int next.|x:4|))
               next.|x:4|)
            (<
               next.|x:4|
               (+
                  (to_real
                     (to_int next.|x:4|))
                  1.0))))
      (= next.|x:11| next.|x:4|)
      (=
         next.|__no_static|
         (=
            next.|x:11|
            (to_real
               (to_int next.|x:11|))))
      (=
         next.ok2
         (=
            (=
               (to_int
                  (- next.|x:4|))
               (-
                  (to_int next.|x:4|)))
            next.|__no_static|))
      (=
         next.ok3
         (=
            (ite
               (< 5 0)
               (-
                  (to_int
                     (/
                        (to_real next.y)
                        (to_real
                           (- 5)))))
               (to_int
                  (/
                     (to_real next.y)
                     (to_real 5))))
            (to_int
               (/
                  (to_real next.y)
                  5.0))))
      (=
         next.ok4
         (=
            (ite
               (< 5 0)
               (-
                  (to_int
                     (/
                        (to_real
                           (to_int next.|x:4|))
                        (to_real
                           (- 5)))))
               (to_int
                  (/
                     (to_real
                        (to_int next.|x:4|))
                     (to_real 5))))
            (to_int
               (/ next.|x:4| 5.0))))))

(query
   TS
   (or |gal-initializing| ok4))
(query
   TS
   (or |gal-initializing| ok3))
(query
   TS
   (or |gal-initializing| ok2))
(query
   TS
   (or |gal-initializing| ok1))

[Lustre] Validating properties:
[Lustre] Property ok4... considering simultaneous states to depth 0
considering simultaneous states to depth 0
[[1;92mValid[0m] considering simultaneous states to depth 0
[Lustre] Property ok3... considering simultaneous states to depth 0
considering simultaneous states to depth 0
[[1;92mValid[0m] considering simultaneous states to depth 0
[Lustre] Property ok2... considering simultaneous states to depth 0
considering simultaneous states to depth 0
[[1;92mValid[0m] considering simultaneous states to depth 0
[Lustre] Property ok1... considering simultaneous states to depth 0
considering simultaneous states to depth 0
[[1;92mValid[0m] considering simultaneous states to depth 0
[Lustre] Model status: [[1;92mValid[0m] 
